### PR TITLE
Add Here Provider

### PR DESCRIPTION
**Pull request by sheub**
Happy to contribute with the Here provider.

### CODE REVIEW

File: src/Provider/Here/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Change Log
+
+The change log describes what is "Added", "Removed", "Changed" or "Fixed" between each release.
+
+## 4.0.0
```
Comment (Nyholm) [inline src/Provider/Here/CHANGELOG.md]: Lets do 1.0.0 for a new provider.

File: src/Provider/Here/Here.php [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+     */
+    private $appId = null;
+
+    /**
+     * @var string
+     */
+    private $appCode = null;
+
+    /**
+     * @param HttpAdapterInterface $adapter An HTTP adapter.
```
Comment (Nyholm) [inline src/Provider/Here/Here.php]: This is wrong. This should be `HttpClient`

**Comment by Mechazawa** [comment]
Thanks for making a PR @sheub! I've been needing this provider support for a while.

**Comment by sheub** [comment]
Hi Nyholm,

I have added the Test Files and could correct some errors, which is good. Nevertheless, despite my efforts I did not manage to get an error free result (see here after for some more details)...

Two small remarks:

1. I had to add  `protected function getApiKey()`  (which won't be used) to the class `class IntegrationTest extends ProviderIntegrationTest` because I got an error: _abstract method must be declared abstract or implement the remaining methods_

2. I don't know what I should do with the issue from styleci 
-        $json = <<<JSON
+        $json = <<<'JSON'


Please let me know 

Sheub





The tests gave:

There was 1 error:

1) Geocoder\Provider\Here\Tests\IntegrationTest::testGeocodeIpv4
Geocoder\Exception\UnsupportedOperation: The Here provider does not support IP addresses, only street addresses.

D:\Code\Geocoder-master\Geocoder\src\Provider\Here\Here.php:75
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\vendor\geocoder-php\provider-integration-tests\src\ProviderIntegrationTest.php:191

--

There were 6 failures:

1) Geocoder\Provider\Here\Tests\HereTest::testReverseWithRealCoordinatesReturnsSingleResult
Failed asserting that 2.38948 matches expected 2.38883.

D:\Code\Geocoder-master\Geocoder\src\Provider\Here\Tests\HereTest.php:153

2) Geocoder\Provider\Here\Tests\IntegrationTest::testExceptions with data set #10 (Geocoder\Query\GeocodeQuery Object (...), 'Geocoder\Exception\InvalidSer...sponse', Nyholm\Psr7\Response Object (...), 'Server 500')
Failed asserting that exception of type "Geocoder\Exception\UnsupportedOperation" matches expected exception "Geocoder\Exception\InvalidServerResponse". Message was: "The Here provider does not support IP addresses, only street addresses." at
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\Here.php:75
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\vendor\geocoder-php\provider-integration-tests\src\ProviderIntegrationTest.php:234
.

3) Geocoder\Provider\Here\Tests\IntegrationTest::testExceptions with data set #11 (Geocoder\Query\GeocodeQuery Object (...), 'Geocoder\Exception\InvalidSer...sponse', Nyholm\Psr7\Response Object (...), 'Server 400')
Failed asserting that exception of type "Geocoder\Exception\UnsupportedOperation" matches expected exception "Geocoder\Exception\InvalidServerResponse". Message was: "The Here provider does not support IP addresses, only street addresses." at
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\Here.php:75
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\vendor\geocoder-php\provider-integration-tests\src\ProviderIntegrationTest.php:234
.

4) Geocoder\Provider\Here\Tests\IntegrationTest::testExceptions with data set #12 (Geocoder\Query\GeocodeQuery Object (...), 'Geocoder\Exception\InvalidSer...sponse', Nyholm\Psr7\Response Object (...), 'Empty response')
Failed asserting that exception of type "Geocoder\Exception\UnsupportedOperation" matches expected exception "Geocoder\Exception\InvalidServerResponse". Message was: "The Here provider does not support IP addresses, only street addresses." at
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\Here.php:75
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\vendor\geocoder-php\provider-integration-tests\src\ProviderIntegrationTest.php:234
.

5) Geocoder\Provider\Here\Tests\IntegrationTest::testExceptions with data set #13 (Geocoder\Query\GeocodeQuery Object (...), 'Geocoder\Exception\InvalidCredentials', Nyholm\Psr7\Response Object (...), 'Invalid credentials response')
Failed asserting that exception of type "Geocoder\Exception\UnsupportedOperation" matches expected exception "Geocoder\Exception\InvalidCredentials". Message was: "The Here provider does not support IP addresses, only street addresses." at
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\Here.php:75
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\vendor\geocoder-php\provider-integration-tests\src\ProviderIntegrationTest.php:234
.

6) Geocoder\Provider\Here\Tests\IntegrationTest::testExceptions with data set #14 (Geocoder\Query\GeocodeQuery Object (...), 'Geocoder\Exception\QuotaExceeded', Nyholm\Psr7\Response Object (...), 'Quota exceeded response')
Failed asserting that exception of type "Geocoder\Exception\UnsupportedOperation" matches expected exception "Geocoder\Exception\QuotaExceeded". Message was: "The Here provider does not support IP addresses, only street addresses." at
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\Here.php:75
D:\Code\Geocoder-master\Geocoder\src\Provider\Here\vendor\geocoder-php\provider-integration-tests\src\ProviderIntegrationTest.php:234

Tests: 30, Assertions: 124, Errors: 1, Failures: 6, Skipped: 1

File: src/Provider/Here/Here.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+     * @var string
+     */
+    private $appCode = null;
+
+    /**
+     * @param HttpClient $adapter An HTTP adapter.
+     * @param string     $appId   An App ID.
+     * @param string     $apoCode An App code.
+     */
+    public function __construct(HttpClient $client, $appId, $appCode)
```
Comment (Baachi) [inline src/Provider/Here/Here.php]: I would typehint the `$appId` and `appCode` parameters as `string`. Than we can remove the exception in the constructor.

File: src/Provider/Here/LICENSE [hunk 4]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2011 — William Durand <william.durand1@gmail.com>
```
Comment (Baachi) [inline src/Provider/Here/LICENSE]: Maybe you want to change it to your name?

File: src/Provider/Here/Tests/HereTest.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    /**
+     * @expectedException \Geocoder\Exception\InvalidCredentials
+     * @expectedExceptionMessage Invalid or missing api key.
+     */
+    public function testGeocodeInvalidApiKey()
+    {
+        $provider = new Here(
+            $this->getMockedHttpClient(
+                '{
+					    "type": {
```
Comment (Baachi) [inline src/Provider/Here/Tests/HereTest.php]: You should use spaces instead of tabs.

File: src/Provider/Here/Tests/IntegrationTest.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+namespace Geocoder\Provider\Here\Tests;
+
+use Geocoder\IntegrationTest\ProviderIntegrationTest;
+use Geocoder\Provider\Here\Here;
+use Http\Client\HttpClient;
+
+/**
+ * @author Sébastien Barré <sebastien@sheub.eu>
```
Comment (Baachi) [inline src/Provider/Here/Tests/IntegrationTest.php]: In the `Here` class you annotate the class with [info@zoestha.de](https://github.com/geocoder-php/Geocoder/pull/844/files#diff-fa50f87c5dfdc2fde6b82c55626dce1aR28) and here with `sebastien@sheub.eu`. Is this intended or do you want to fix this?

**Comment by sheub** [comment]
Hi there, 
the errors from Travis are : 
Geocoder\Exception\InvalidCredentials.
Should I use real credentials also for the Travis build ?

**Comment by Nyholm** [comment]
You should not commit the credentials. But you should commit all the files created in `Tests/.cached_responses`.

**Comment by Nyholm** [comment]
Thank you. 
You should add an entry for `HERE_APP_ID` in phpunit.xml.dist. See https://github.com/geocoder-php/Geocoder/blob/master/phpunit.xml.dist#L27

File: src/Provider/Here/Tests/IntegrationTest.php (hunk 6 above)
Comment (sheub) [reply src/Provider/Here/Tests/IntegrationTest.php]: Thanks for the Review. I have corrected the Here Class.

**Comment by sheub** [comment]
Hey there,
still get InvalidCredentials.
please let me know what should be done before the merge.

**Comment by sheub** [comment]
Can anyone give me instructions on how to finalize this PR? In advance Thank you.

**Comment by Baachi** [comment]
Hey @sheub,

the TravisCI checks must be passed. A failing PR can't get merged.

**Comment by sheub** [comment]
Hey @Baachi,
Thanks for your reaction. 

The errors from Travis are :
Geocoder\Exception\InvalidCredentials.

But he integration tests passed when I run them at home with my credentials.

I have commit all the files created in Tests/.cached_responses and I have added
```
<server name="HERE_APP_ID" value="YOUR_APP_ID" />
<server name="HERE_APP_CODE" value="YOUR_APP_CODE" />
```
in phpunit.xml.dist



So I don't understand what I am missing or what I am doing wrong. Thanks for your support.

**Comment by sheub** [comment]
Hi, it seems that the files created in the ./cached_responses will get a different Hash code for each `HERE_APP_ID`/`HERE_APP_CODE` pair.

see:
`geocoder.api.here.com_62eb9554b9240ece07c143250a367bf2a534ac37`
and
`geocoder.api.here.com_194f5df1c578dfb8b688908c8e87c11d2afdcf9d`
if I run with a real credentials or with `YOUR_APP_ID` `YOUR_APP_CODE`

For that reason the tests is not run on the ./cached_responses.

Any Idea? Thanks for your support.

**Comment by sheub** [comment]
Hi @Nyholm ,
May I kindly ask you to point me out in the right direction so I can resolve this issue.
I can't see why the tests won't run on the ./cached_responses
Many thanks for your time.

**Comment by sheub** [comment]
Hi @Nyholm ,

I could find out why the tests wouldn't run on the cached files. 
The problem is that the HERE provider requires 2 Credential keys in the url: the `app_id` & the `app_code`.

and only one apiKey is searched for and replaced in the body part of the cacheKey:
Geocoder\IntegrationTest\CachedResponseClient
```php
     public function sendRequest(RequestInterface $request)
    {
        $url = (string) $request->getUri();
        $host = (string) $request->getUri()->getHost();
        if (!empty($this->apiKey)) {
            $url = str_replace($this->apiKey, '[apikey]', $url);
        }

        $file = sprintf('%s/%s_%s', $this->cacheDir, $host, sha1($url));
        if (is_file($file) && is_readable($file)) {
            return new Response(200, [], (new StreamFactory())->createStream(unserialize(file_get_contents($file))));
        }

        $response = $this->delegate->sendRequest($request);
        file_put_contents($file, serialize($response->getBody()->getContents()));

        return $response;
    }
```

I could get it work properly at home, but I had to (slightly) modify three files from `geocoder-php/provider-integration-tests`:

- Geocoder\IntegrationTest\ProviderIntegrationTest, 
- Geocoder\IntegrationTest\CachedResponseClient
- Geocoder\IntegrationTest\BaseTestCase

In a way that both the  `app_id` & the `app_code` are replaced in the url.

This solution is working, but it requires that the `geocoder-php/provider-integration-tests` will be modified...

**Comment by jbelien** [comment]
Hello @sheub ,

I didn't check in details but Google Maps provider also requires 2 parameters (ClientId and APIKey) ; so I guess it should also work with HERE without having to modify the `Geocoder\IntegrationTest`.

Have a look at https://github.com/geocoder-php/Geocoder/blob/master/src/Provider/GoogleMaps/

**Comment by sheub** [comment]
I'll have a look, but I am using the Google Maps provider with only one key (GOOGLEMAP_API_KEY)

**Comment by sheub** [comment]
Maybe someone could tell me if it is **for sure** possible to get the caching to work with the 2 Credential keys, before I send some more hours on it. Thank you.

File: src/Provider/Here/Here.php [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * @var string
+     */
+    private $appId = null;
+
+    /**
+     * @var string
+     */
+    private $appCode = null;
```
Comment (Nyholm) [inline src/Provider/Here/Here.php]: You do not need the `= null`

File: src/Provider/Here/Readme.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+This is the Here provider from the PHP Geocoder. This is a **READ ONLY** repository. See the
+[main repo](https://github.com/geocoder-php/Geocoder) for information and documentation. 
+
+### Install
+
+```bash
+composer require geocoder-php/here-provider
+```
+### Note
```
Comment (Nyholm) [inline src/Provider/Here/Readme.md]: This line should be removed.

File: src/Provider/Here/Readme.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+## App Id and App Code
```
Comment (Nyholm) [inline src/Provider/Here/Readme.md]: Use `###` instead of `##`

File: src/Provider/Here/Readme.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+Get your Here credentials at https://developer.here.com/
+
+## Language parameter
```
Comment (Nyholm) [inline src/Provider/Here/Readme.md]: Use `###` instead of `##`

File: src/Provider/Here/Tests/HereCachedResponseClient.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+use Nyholm\Psr7\Factory\StreamFactory;
+use Nyholm\Psr7\Response;
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Serve responses from local file cache.
+ *
+ * @author Sébastien Barré <sebastien@sheub.eu> override the provider-integration-tests: CachedResponseClient class
+ */
+class HereCachedResponseClient extends CachedResponseClient
```
Comment (Nyholm) [inline src/Provider/Here/Tests/HereCachedResponseClient.php]: Good

Comment (Nyholm) [reply src/Provider/Here/Tests/HereCachedResponseClient.php]: This is the best solution I could think of. I suggest to modify the `CachedResponseClient` only if other providers also need this functionality.

**Comment by Nyholm** [comment]
I added some notes. These are just minors. I will merge this PR now and prepare the package.

**Comment by Nyholm** [comment]
The package is added: https://packagist.org/packages/geocoder-php/here-provider

You could improve your tests now when 1.1.0 is released of the integration tests. That version includes your PR. 

Let me know when I should tag version 1.0.0 of this new package. (Ie after you test it a little bit)

**Pull request merged**